### PR TITLE
Apply clang-tidy to runtime capi directory

### LINKED ISSUE #1271: Continue applying clang-tidy fixes to Catalyst headers fetched from Lightning
#### Issue description

Description of the issue - When moving `Test_LightningDriver.cpp` from Catalyst to Lightning (https://github.com/PennyLaneAI/pennylane-lightning/pull/960), Lightning will need to fetch new headers:

- `ExecutionContext.hpp`
- `Python.hpp`
- `TestUtils.hpp`

Which at the moment do not pass the `clang-tidy` checks performed at the Lightning side.

* *Expected behavior:* `clang-tidy` checks passed.

* *Actual behavior:* `clang-tidy` checks not passed (see log here: https://github.com/PennyLaneAI/pennylane-lightning/actions/runs/11649901515/job/32437990696?pr=960)

* *Reproduces how often:* 100%

### PR DESCRIPTION
**Pull request by paul0403**
**Context:**
We keep applying clang-tidy suggestions to the files in `runtime`. 

**Description of the Change:**
Apply clang-tidy to runtime capi directory.

**Benefits:**
Better following of clang rules; unblocks the moving of lightning files from catalyst repo to lightning repo.

**Related GitHub Issues:** closes #1271 
[sc-77391]

### CODE REVIEW

File: runtime/CMakeLists.txt
Comment (paul0403) [inline runtime/CMakeLists.txt:17]: restore before merging

File: runtime/Makefile
Comment (paul0403) [inline runtime/Makefile:14]: restore before merging

File: runtime/Makefile
Comment (paul0403) [inline runtime/Makefile:109]: restore before merging

**Review by rauletorresc — APPROVED** [review]
LGTM, let me know what is left to refactor

File: runtime/lib/capi/ExecutionContext.hpp
Comment (maliasadi) [inline runtime/lib/capi/ExecutionContext.hpp:334]: Was there any warning for this? :eyes:

File: runtime/lib/capi/ExecutionContext.hpp
Comment (maliasadi) [inline runtime/lib/capi/ExecutionContext.hpp:1]: Could you also run tidy for the include folder? I saw a few warnings with [DataView](https://github.com/PennyLaneAI/catalyst/blob/main/runtime/include/DataView.hpp) the last time I locally checked.

File: runtime/lib/capi/ExecutionContext.hpp
Comment (paul0403) [reply runtime/lib/capi/ExecutionContext.hpp:355]: Yes, clang-tidy really wants all comparisons to be as explicit as possible

File: runtime/lib/capi/ExecutionContext.hpp
Comment (rauletorresc) [reply runtime/lib/capi/ExecutionContext.hpp:1]: Those were tackled here, as far as i know: https://github.com/PennyLaneAI/catalyst/pull/1261/files#diff-3ea0859d40991b00b11aa0d1cf6175c4adb2d06064cce30716349e84f00d58eeL102

Comment (paul0403) [reply runtime/lib/capi/ExecutionContext.hpp:1]: dataview was already tidied in #1261

File: runtime/lib/capi/ExecutionContext.hpp
Comment (rauletorresc) [reply runtime/lib/capi/ExecutionContext.hpp:355]: Not lately. Last ones reported were:
```
/home/runner/work/pennylane-lightning/pennylane-lightning/Build/pennylane_lightning/core/src/simulators/lightning_qubit/catalyst/../../include/ExecutionContext.hpp:41:7: error: class 'MemoryManager' defines a non-default destructor but does not define a copy constructor, a copy assignment operator, a move constructor or a move assignment operator [cppcoreguidelines-special-member-functions,hicpp-special-member-functions,-warnings-as-errors]
class MemoryManager final {
      ^
/home/runner/work/pennylane-lightning/pennylane-lightning/Build/pennylane_lightning/core/src/simulators/lightning_qubit/catalyst/../../include/ExecutionContext.hpp:54:13: error: do not manage memory manually; use RAII [hicpp-no-malloc,-warnings-as-errors]
            free(allocation);
```

Comment (paul0403) [reply runtime/lib/capi/ExecutionContext.hpp:355]: @rauletorresc Converted to smart pointers https://github.com/PennyLaneAI/catalyst/pull/1273/commits/de8e1ba30b5752ff4123f012ba306fa22b42db2a

**Review by rauletorresc — CHANGES_REQUESTED** [review]
These ones are still failing:
```
/home/runner/work/pennylane-lightning/pennylane-lightning/Build/pennylane_lightning/core/src/simulators/lightning_qubit/catalyst/../../include/ExecutionContext.hpp:51:33: error: do not manage memory manually; consider a container or a smart pointer [hicpp-no-malloc,-warnings-as-errors]
        std::shared_ptr<void> p(std::malloc(size), [](void *ptr) { free(ptr); });
                                ^~~~~~~~~~~~~~~~~
/home/runner/work/pennylane-lightning/pennylane-lightning/Build/pennylane_lightning/core/src/simulators/lightning_qubit/catalyst/../../include/ExecutionContext.hpp:51:68: error: do not manage memory manually; use RAII [hicpp-no-malloc,-warnings-as-errors]
        std::shared_ptr<void> p(std::malloc(size), [](void *ptr) { free(ptr); });
                                                                   ^~~~~~~~~
/home/runner/work/pennylane-lightning/pennylane-lightning/Build/pennylane_lightning/core/src/simulators/lightning_qubit/catalyst/../../include/ExecutionContext.hpp:58:86: error: do not manage memory manually; use RAII [hicpp-no-malloc,-warnings-as-errors]
        std::shared_ptr<void> p(std::aligned_alloc(alignment, size), [](void *ptr) { free(ptr); });
                                                                                     ^~~~~~~~~
/home/runner/work/pennylane-lightning/pennylane-lightning/Build/pennylane_lightning/core/src/simulators/lightning_qubit/catalyst/../../include/ExecutionContext.hpp:63:39: error: the parameter 'ptr' is copied for each invocation but only used as a const reference; consider making it a const reference [performance-unnecessary-value-param,-warnings-as-errors]
    void insert(std::shared_ptr<void> ptr)
                                      ^
                const                &
/home/runner/work/pennylane-lightning/pennylane-lightning/Build/pennylane_lightning/core/src/simulators/lightning_qubit/catalyst/../../include/ExecutionContext.hpp:76:36: error: loop variable is copied but only used as const reference; consider making it a const reference [performance-for-range-copy,-warnings-as-errors]
        for (std::shared_ptr<void> sharedP : _impl) {
                                   ^
             const                &
/home/runner/work/pennylane-lightning/pennylane-lightning/Build/pennylane_lightning/core/src/simulators/lightning_qubit/catalyst/../../include/ExecutionContext.hpp:90:36: error: loop variable is copied but only used as const reference; consider making it a const reference [performance-for-range-copy,-warnings-as-errors]
        for (std::shared_ptr<void> sharedP : _impl) {
                                   ^
             const                &
```

**Comment by paul0403** [comment]
Discussed offline, we will skip smart pointers here to quickly unblock lightning downstreaming, and do smart pointers in a separate PR #1284

**Comment by paul0403** [comment]
@rauletorresc I think https://github.com/PennyLaneAI/catalyst/pull/1273/commits/95facd7f86e430aec520bf1a8c6d515b7828e01c is good to go. Can you run tidy in your Lightning PR to make sure?

**Review by rauletorresc — APPROVED** [review]

**Pull request merged**
